### ISSUE TITLE
master deployment

### ISSUE THREAD

**Issue by markus2330** [ElektraInitiative/PermaplanT#37]
- [x] proxy setup for backend
- [x] Ansible setup to install PostgreSQL etc. (is already installed)
- [x] build&deploy [ci/Jenkinsfile](https://github.com/ElektraInitiative/PermaplanT/blob/master/ci/Jenkinsfile) (cargo run, diesel setup, diesel migration run)
- [x] make Jenkins job public (or at least accessible for PermaplanT users)
- [ ] (database) backup

**Comment by markus2330** [comment]
@aidnurs @buenaflor How will we do the proxy-setup so that both the Frontend and Backend are reachable on dev.permaplant.net?

**Comment by lukashartl** [comment]
Hello,

short update on the infrastructure work :rocket:

The following things have now been automized (with ansible):

* install postgres
* create postgres database + database user
* set database password
* write `.env` file
* create service user (the service will run as a user with minimal privileges)

Next up will be the [ci/Jenkinsfile](https://github.com/ElektraInitiative/PermaplanT/blob/master/ci/Jenkinsfile) as well as the whole apache2 + proxy + monit.

Expect more updates soon :rocket:

---

> @aidnurs @buenaflor How will we do the proxy-setup so that both the Frontend and Backend are reachable on dev.permaplant.net?

The following two solutions come to my mind:

1. We create an `api.dev.permaplant.net` which will only host the backend and `dev.permaplant.net` which only hosts the static (css+js) files.
2. We utilize subpaths. The Location `dev.permaplant.net/api` gets proxied to the backend, while `dev.permaplant.net` just holds the static files.

Either way, we have to inject the correct BASE_URL for backend requests into the frontend. This could happen via environment variables during build time, etc. but maybe @aidnurs @buenaflor have any better ideas :smiley:

**Comment by markus2330** [comment]
Great job! :1st_place_medal: Looking forward to the PR! 🚀

**Comment by lukashartl** [comment]
Hello,

next update on the infrastructure :rocket:

Dev uses the latest rust version via rustup installed to `/opt/rust`. A `source`-able env script is provided under `/opt/rust.env`. Because the `rust` version provided by the os package manager is too old (of course).

I've tried to build the latest master on dev, but unfortunately I got errors like:

```
error: Table contains more than 32 columns. Consider enabling the `64-column-tables` feature to enable diesels support for tables with more than 32 columns.
error[E0433]: failed to resolve: use of undeclared crate or module `plant_detail`
```

@aidnurs @buenaflor Do you have any idea on what I could have done wrong?

`backend/README.md` states the following commands:

```
cargo install diesel_cli --no-default-features --features postgres
LC_ALL=C ~/.cargo/bin/diesel setup # needs db
LC_ALL=C ~/.cargo/bin/diesel migration run # needs db
cargo install typeshare-cli
cargo build
cargo run
```

Unfortunately I need to issue the `diesel` commands for the schema.rs to be written, but `diesel` already requires a functioning Postgres database. In the build phase I don't want to connect to any database. Connecting to the database is ok for testing and for the development server, but not in the build step as it should just throw out the binary.

Questions for the meeting on 2022-02-17:

1. Should we pin the rust version in any way, or is it ok to always use the latest version?
2. How should we handle the situation during build? (no Postgres, etc.)
3. How do we inject the correct `BASE_URL` for the backend into the frontend?

**Comment by markus2330** [comment]
> next update on the infrastructure :rocket:

Great to see so many&fast updates!

> Dev uses the latest rust version via rustup

Perfect. As written in the docu it should be the stable channel and fixed with 2023-02-09, rust version 1.67.1 (current version).

> I've tried to build the latest master on dev, but unfortunately I got errors like:

Please don't hesitate to open new issues, especially if you simply did what is written in the tutorials because then it is a problem either way (either functionality is broken or the tutorial is wrong).

Should be fixed in #52.

> Unfortunately I need to issue the diesel commands for the schema.rs to be written, but diesel already requires a functioning Postgres database.

Yes, because diesel also changes the Postgres database (migrates its tables to the latest version). So this is required per design even in production.

See also #53

> How should we handle the situation during build? (no Postgres, etc.)

See #50 (with my review)

> How do we inject the correct BASE_URL for the backend into the frontend?

See #49.

**Comment by lukashartl** [comment]
Hello,

time for another update on the infrastructure work :)

I've created PR #54 which includes a docker-compose file. This files separates everything into containers and only shares the bindings between frontend and backend via a volume.
It doesn't use a volume for the postgres database, so we can let this run for each PR and it will always run all migrations.

Later on, we could also import the latest dev database and try to migrate it up as well (so we have both cases: new database and migrating an already existing one).

I think we can also use this in dev/production, if we actually give a volume to the postgres container and make all ports accessible and configurable (via the environment). Of course we shouldn't use the npm dev server in dev/prod!

The next step would be to write the script that should be called inside the jenkins pipeline as well as setting up dev.

/cc @markus2330 could you check this out (especially the split between build-time (docker) and run-time (entry points) in rust)

**Comment by markus2330** [comment]
Thanks for the update!

I am a bit confused :confused:, the last time we talked it was about scripts directly running in LXC:

1. Ansible sets up Postgres, Proxy etc.
2. CI basically only locally compiles (which already modifies the DB) and then copies files to the right places.

In the current LXC setup Docker won't work (nesting permissions not given).

This doesn't mean that we shouldn't have Docker, definitely for tests, maybe for dev or other kinds of deployments. But I wouldn't focus on that right now.

Btw. I didn't see a PR for Ansible.

**Comment by lukashartl** [comment]
I know, but we talked about running a full compile etc. (not only tests) during the last meeting within separated containers.

So I thought if we utilize docker right know, we could use it for the build pipeline (e.g. testing if code changes work), to run dev env and to run production env.

Should we go back to running directly on the servers without Docker? Wouldn't be a hassle at all, since most of the work I put in was about separating build time and run time (as I have not used rust before) :)
So going "back" to running it directly on the server is absolutely no problem 😄
(I'd still keep the docker-compose files in the PR, as we might need it later)

**Comment by markus2330** [comment]
Yes, please let us keep things as simple as possible for the first version.

Please create a separate PR for the CI scripts, as I don't want to spent time reviewing docker-compose before the first milestone (unless we actually need it).

**Comment by lukashartl** [comment]
The Jenkins Job is now readable for the public :rocket:

**Comment by lukashartl** [comment]
The dev environment is up now :rocket:

```
$ curl https://dev.permaplant.net/api/plants
{"message":"ok","data":[{"id":1,"tags":["Fruit crops"],"species":"Wildtomate","plant":null,"plant_type":null}]}
```

We will auto-deploy master after #78 is merged ☺️ 

P.S. I haven't written `sysvinit`-style scripts for quite a few years now 😸

**Comment by markus2330** [comment]
I moved some tasks to #79.

> curl https://dev.permaplant.net/api/plants

:fireworks: 

> We will auto-deploy master after https://github.com/ElektraInitiative/PermaplanT/pull/78 is merged relaxed

Perfect. Is the PR deployment still functional afterwards?

> P.S. I haven't written sysvinit-style scripts for quite a few years now smile_cat

Why not simply throw it in inittab with respawn? Actually I thought that monit has a similar feature (respawn on processes that exit) but when searching for it I did not find it. :see_no_evil:

**Comment by lukashartl** [comment]
> Perfect. Is the PR deployment still functional afterwards?

If you mean testing each PR, I haven't touched that part at all. So yes, it is still functional :)

> Why not simply throw it in inittab with respawn? Actually I thought that monit has a similar feature (respawn on processes that exit) but when searching for it I did not find it. 🙈

IIRC I wouldn't have fine control over the service (e.g. ad-hoc start/stop/status), and during the deployment I need to stop, move the binary, and restart the service.

I'd just use monit to monitor and restart the service, but not as a service manager.

**Comment by markus2330** [comment]
> If you mean testing each PR, I haven't touched that part at all. So yes, it is still functional :)

And deploying PRs? Please make clear in the documentation what already works and what is to be done.

> IIRC I wouldn't have fine control over the service (e.g. ad-hoc start/stop/status), and during the deployment I need to stop, move the binary, and restart the service.

This shouldn't be needed, you can simply overwrite the binary and then kill the process. Later we don't even restart as we want to implement hot reloading (by your buddy).

> I'd just use monit to monitor and restart the service, but not as a service manager.

It doesn't really answer the question why inittab isn't used. But let us leave it as it is now and re-investigate when we actually have hot reloading.

**Comment by lukashartl** [comment]
@markus2330 can you please setup a DNS record for pr.permaplant.net like you did for dev.permaplant.net. I'll take care of proxy, letsencrypt and everything else.

**Comment by markus2330** [comment]
pr.permaplant.net was added.

I also created git.permaplant.net and issues.permaplant.net, which should later redirect to the GitHub repo and its issues tracker.

> I'll take care of proxy, letsencrypt and everything else.

Thank you so much :heart:

**Comment by markus2330** [comment]
Btw. actually #53 is the issue about pr.permaplant.net

Btw2. I moved "testing" to #98.
